### PR TITLE
add(internal): rules feature

### PR DESCRIPTION
**Pull request by titpetric**
- add make test.rules target
- add rules api implementation
- add sql migration for rules table
- add unit tests to verify it works

### CODE REVIEW

File: system/db/schema/mysql/20190116102104.rules.up.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE `sys_rules` (
+  `rel_team` BIGINT UNSIGNED NOT NULL,
+  `resource` VARCHAR(128) NOT NULL,
+  `operation` VARCHAR(128) NOT NULL,
+  `value` ENUM('no', 'yes') NOT NULL,
```
Comment (darh) [inline system/db/schema/mysql/20190116102104.rules.up.sql]: change to `boolean` or `tinyint` for consistency

File: internal/rules/resources.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		// select rules
+		"select r.value from sys_rules r",
+		// join members
+		"inner join sys_team_member m on (m.rel_team = r.rel_team and m.rel_user=?)",
+		// add conditions
+		"where r.resource LIKE ? and r.operation=?",
+	}
+	resource = strings.Replace(resource, "*", "%", -1)
+	queryString := strings.Join(query, " ")
+	if err := r.db.Select(&result, queryString, user, resource, operation); err != nil {
```
Comment (darh) [inline internal/rules/resources.go]: should be in repository pkg

File: internal/rules/resources_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+			i++
+		}
+		db.Insert("sys_team_member", types.TeamMember{TeamID: 1, UserID: user.ID})
+		db.Insert("sys_team_member", types.TeamMember{TeamID: 2, UserID: user.ID})
+
+		resources := rules.NewResources(ctx, db)
+
+		// default (unset=deny)
+		{
+			mustFail(t, resources.CheckAccess("channel:1", "edit"))
```
Comment (darh) [inline internal/rules/resources_test.go]: why not split this into resource type & resource id?

File: internal/rules/resources.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	queryString := strings.Join(query, " ")
+	if err := r.db.Select(&result, queryString, user, resource, operation); err != nil {
+		return err
+	}
+
+	// order by deny, allow
+	for _, val := range result {
+		if val == Deny {
+			return errors.New("Access not allowed")
```
Comment (darh) [inline internal/rules/resources.go]: is this really an error? shouldn't this return bool only?

File: internal/rules/resources_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	"testing"
+
+	"github.com/titpetric/factory"
+
+	"github.com/crusttech/crust/internal/auth"
+	"github.com/crusttech/crust/internal/rules"
+	"github.com/crusttech/crust/system/types"
+)
+
+func TestRules(t *testing.T) {
```
Comment (darh) [inline internal/rules/resources_test.go]: I do not see the point in testing if the DB works, should be more like service/repo struct we have elsewhere w/  proper unit testing.

File: internal/rules/main_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	// migrate database schema
+	if err := systemMigrate.Migrate(db); err != nil {
+		log.Printf("Error running migrations: %+v\n", err)
+		return
+	}
+
+	os.Exit(m.Run())
+}
+
+func assert(t *testing.T, ok bool, format string, args ...interface{}) bool {
```
Comment (darh) [inline internal/rules/main_test.go]: I've started with abstracting this `assert` code repetition into `internal/test.Assert`

File: internal/rules/resources.go (hunk 2 above)
Comment (titpetric) [reply internal/rules/resources.go]: shouldn't, code is shared between messaging and crm

File: internal/rules/resources_test.go (hunk 3 above)
Comment (titpetric) [reply internal/rules/resources_test.go]: Mostly to have a unique resource id that identifies the type as well as the record (it comes from daasi api-s as well so we need to be somewhat compatible here). As for how the record ID is put together, there's a common function available for record entries under sam/types currently (ResourceID() or something like that).

File: internal/rules/resources.go (hunk 4 above)
Comment (titpetric) [reply internal/rules/resources.go]: Yes and no. Errors are easier to handle, but a TODO which you can see here is to add error types so we can check higher up if the returned error is a ErrDeniedAccess or something else (database issue, or with didmos a http response issue)...

tl;dr: introduce error type for ErrDeniedAccess; and maybe (@mitjaziv opinion) a helper function that just logs the error and returns a bool?

File: internal/rules/resources_test.go (hunk 5 above)
Comment (titpetric) [reply internal/rules/resources_test.go]: 1. more tests > less tests
2. real environment > mocks

Most of the tests in internal/ (at least rbac and rules) are to verify functionality against real services, where mocking won't tell you if what actually is implemented works as it should. These aren't tests for the CI (but it would be obviously bad if they would break).

Mock tests will be added once the interface is finalized. I'm not 100% happy with the current interface as it can be simplified further (no need for Multi, there's a stutter with CheckAccess vs. just Check,...), but that's neither here nor there.

File: internal/rules/main_test.go (hunk 6 above)
Comment (titpetric) [reply internal/rules/main_test.go]: Noted, TODO for a cleanup sprint. A few suggestions:

1. rename ErrNil into NoError, - previously `must()`
2. add Error(), - previously `mustFail()`

I looked at [testify](https://godoc.org/github.com/stretchr/testify/require) for the naming here.

File: internal/rules/resources_test.go (hunk 5 above)
Comment (titpetric) [reply internal/rules/resources_test.go]: I think this will be a learning moment for me: show me how ;)

File: internal/rules/main_test.go (hunk 6 above)
Comment (titpetric) [reply internal/rules/main_test.go]: If we can find something better with a smaller footprint sure. Maybe I'm being too pedantic, but 3 functions vs. a package with ~100 of them... I'll leave it to you, since you started with this, so +1 if you want ;)

File: internal/rules/resources_test.go (hunk 5 above)
Comment (darh) [reply internal/rules/resources_test.go]: /slow/!short, sorry

```
	if testing.Short() {
		log.Println("skipping test in short mode.")
		return
	}
```

Comment (titpetric) [reply internal/rules/resources_test.go]: Already done in main_test.go, I'll add the log info

File: internal/rules/resources.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+	"github.com/crusttech/crust/internal/auth"
+	"github.com/crusttech/crust/system/types"
+)
+
+type Access string
+
+const (
+	Allow   Access = "yes"
+	Deny           = "no"
```
Comment (mitjaziv) [inline internal/rules/resources.go]: Can I suggest usage of true/false instead of yes/no?

Comment (darh) [reply internal/rules/resources.go]: Not sure how this will affect the "inherit" access flag.

@titpetric already confirmed the change on db level: https://github.com/crusttech/crust/pull/20#pullrequestreview-193193169

Comment (mitjaziv) [reply internal/rules/resources.go]: It does not affect inherit flag.

File: internal/rules/resources.go [hunk 8]
```diff
@@ -104,8 +104,9 @@ func (r *resources) Grant(resource string, teamID uint64, operations []string, v
 	row := types.Rules{
 		TeamID:   teamID,
 		Resource: resource,
-		Value:    string(value),
+		Value:    int(value),
```
Comment (titpetric) [inline internal/rules/resources.go]: @mitjaziv can you please move Access to the types package here, so we can avoid the type conversion fully

File: internal/rules/resources.go [hunk 9]
```diff
@@ -11,12 +11,12 @@ import (
 	"github.com/crusttech/crust/system/types"
 )
 
-type Access string
+type Access int
 
 const (
```
Comment (titpetric) [inline internal/rules/resources.go]: Provide variables for these types so we can continue to use `rules.Allow` when we use the rules pkg
